feat: add blog post "How to build Web Components with Svelte" (#2745)
mainmatter/mainmatter.com: src/posts/2025-06-25-web-components-with-svelte.md

---
title: "How to build Web Components with Svelte"
authorHandle: paoloricciuti
tags: [svelte]
customCta: "global/svelte-workshops-cta.njk"
bio: "Paolo Ricciuti, Senior Software Engineer"
description: "What is a Web Component, how to build one and, most importantly, how to do it with Svelte for ease of development"
autoOg: true
tagline: <p>Web Components are cool but very difficult to build... can Svelte be of any help here?</p>

image: "/assets/images/posts/2025-06-25-web-components-with-svelte/header.jpg"
imageAlt: "The Svelte logo on a gray background picture"
---

When I started learning HTML, I was quite fascinated by how a simple `<marquee>` tag could magically make the text inside of it start to move.

<marquee>Come on! Isn't this cool?</marquee>

> But especially when I started building for the web, there weren't that many cool tags: for instance, `<input type="date" />` was not a thing back then, and the number of people who actually **surfed the web** (gosh, I miss the times when the web had all these cool words... I'll secretly continue to call myself a Web Master until the doom of humanity) explicitly disabling JavaScript was a concern to have.

Reviewer: I feel like the side comment is a bit too long for it to be in the middle of the sentence.
Maybe moving it at the end and repeating the **surf the web** remark could work better?

```suggestion
But especially when I started building for the web, there weren't that many cool tags: for instance, `<input type="date" />` was not a thing back then, and the number of people who actually surfed the web  explicitly disabling JavaScript was a concern to have. **Surfed the web** gosh, I miss the times when the web had all these cool words... I'll secretly continue to call myself a Web Master until the doom of humanity
```

- We don't need to parse out `count` manually... Svelte is doing that for us when we specify the `props` attribute of `customElement`.
- Similarly, to sync between `preSentence` and `pre-sentence`, we just need to add the `attribute` property.

Svelte also helps us with how we can interact with our custom element from the outside world:

```svelte
<script>
	import "./CounterComponent.svelte";

	let counter;
</script>

<counter-component bind:this={counter} count={10} pre-sentence="count"></counter-component>

<button onclick={()=>counter.count = 42}>change counts</button>
<button onclick={()=>counter.preSentence = "count is"}>change sentence</button>
```

As you can see, Svelte created getters and setters for our props so that we can access them without having to rely on `setAttribute`.

Once again, here's a [Svelte playground](https://svelte.dev/playground/e309ec57a14d4c11aa3edf1934fb3670?version=5.34.7) if you want to play around with it. (Unfortunately, based on how the playground doesn't refresh the page and how you can't redefine the same custom element twice, you'll need to refresh the page if you want to make a change to the code 🤷🏻‍♂️)

## Where do we go from here?

> I can already hear a question building up in your mind... what even is the point of all this? Why would I build a Web Component with Svelte if I could just use the Svelte component as a Svelte component? Well, the nice thing about Web Components is that they are effectively framework agnostic... if you build your Svelte project as a library and install it in your React project, you can effectively start using Svelte in your legacy-different-framework project TODAY!

Reviewer: The question I have is: "Where's the catch?"?
I don't want to disregard Web Components completely but they've caveats worth considering.
Also unless something's changed, React used to have problems with binding WebComponent events because AFAIK, WCs require using `addEventListener` instead of assigning a function as a property `wc.onmyevent = () => {}`

- You must implement "waiters" to ensure that Web Components have loaded - otherwise you're going to experience pop-ins as they mount
- They must carefully use the DOM and Window APIs if you care about SSR.
- Optimizing the bundle size could be a hassle. I don't imagine that bundles are going to be optimized well when each Web Component is essentially an entrypoint.

Author: I think React now has decently good support for them but the rest are all valid concerns...i'll add a section on the pitfalls of web components and when it's actually reasonable to use them

```svelte
<script>
	import "./CounterComponent.svelte";

	let counter;
</script>

<counter-component bind:this={counter} count={10} pre-sentence="count"></counter-component>

<button onclick={()=>counter.count = 42}>change counts</button>
<button onclick={()=>counter.preSentence = "count is"}>change sentence</button>
```

As you can see, Svelte created getters and setters for our props so that we can access them without having to rely on `setAttribute`.

Once again, here's a [Svelte playground](https://svelte.dev/playground/e309ec57a14d4c11aa3edf1934fb3670?version=5.34.7) if you want to play around with it. (Unfortunately, based on how the playground doesn't refresh the page and how you can't redefine the same custom element twice, you'll need to refresh the page if you want to make a change to the code 🤷🏻‍♂️)

## So... are Web Components the future?

Well... unfortunately no: as we've seen, Web Components are pretty powerful and they are getting even more powerful with the addition of [Declarative Shadow DOM](https://web.dev/articles/declarative-shadow-dom), but they are not short of pitfalls:

> - They require JavaScript, which means that if you don't build them following a certain design pattern (they should basically be used as a wrapper to add interactivity to an already present and styled DOM), they will "pop" into existence causing layout shift.

Reviewer: I don't quite get the remark, but it's a bit long after reading it I lost track of what was said before.
> (they should basically be used as a wrapper to add interactivity to an already present and styled DOM),
It's safe to remove, I'd say?

```suggestion
- They require JavaScript, which means that if you don't build them following a certain design pattern, they might "pop" into existence as they mount, causing layout shift.
```
